From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fredrik=20S=C3=B6derquist?= <fs@opera.com>
Date: Thu, 8 Oct 2026 11:02:30 -0700
Subject: Fix observer iteration in SVGResourceDocumentContent
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Original change's description:
> Fix observer iteration in SVGResourceDocumentContent
>
> The SVGResourceDocumentObserver::ResourceNotifyFinished() callback can
> cause new observers to be added. Iterate in a copy.
>
> Fixed: 570047270
> Change-Id: Idd45163296244443e3ce7d6b292e78fc39aaee83
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8519167
> Commit-Queue: Philip Rogers <pdr@chromium.org>
> Reviewed-by: Philip Rogers <pdr@chromium.org>
> Auto-Submit: Fredrik Söderquist <fs@opera.com>
> Cr-Commit-Position: refs/heads/main@{#1712579}

(cherry picked from commit 035234e0b8a26edd30bcc93f81efd642d099f720)

Bug: 571476855,570047270
Change-Id: Idd45163296244443e3ce7d6b292e78fc39aaee83
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8541191
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Commit-Queue: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Cr-Commit-Position: refs/branch-heads/7871@{#4437}
Cr-Branched-From: f542126b8c1b3e80104b26bb05ec830bd1206f29-refs/heads/main@{#1639810}

diff --git a/third_party/blink/renderer/core/svg/svg_resource_document_content.cc b/third_party/blink/renderer/core/svg/svg_resource_document_content.cc
index 57a62268519ad108117b0085fa87992b67415cbb..05662b2590e640b1c42ce8d0206bf17db8058e5b 100644
--- a/third_party/blink/renderer/core/svg/svg_resource_document_content.cc
+++ b/third_party/blink/renderer/core/svg/svg_resource_document_content.cc
@@ -211,7 +211,8 @@ void SVGResourceDocumentContent::NotifyObserver(
 }
 
 void SVGResourceDocumentContent::NotifyObservers() {
-  for (auto& observer : observers_) {
+  HeapVector<Member<SVGResourceDocumentObserver>> observer_list(observers_);
+  for (auto& observer : observer_list) {
     observer->ResourceNotifyFinished(this);
   }
 }
@@ -250,7 +251,8 @@ void SVGResourceDocumentContent::ContentChanged() {
   if (inhibit_content_change_) {
     return;
   }
-  for (auto& observer : observers_) {
+  HeapVector<Member<SVGResourceDocumentObserver>> observer_list(observers_);
+  for (auto& observer : observer_list) {
     observer->ResourceContentChanged(this);
   }
 }
